### PR TITLE
Reward tracking

### PR DESCRIPTION
**Pull request by maqi**
## Description

reviewpad:summary

### CODE REVIEW

File: sn_cli/src/bin/subcommands/wallet/wo_wallet.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@ use sn_client::transfers::{
     WatchOnlyWallet,
 };
 use sn_client::Client;
+use sn_transfers::CASH_NOTE_PURPOSE_FOR_TRANSFER;
```
Comment (b-zee) [inline sn_cli/src/bin/subcommands/wallet/wo_wallet.rs:23]:
```suggestion
use sn_transfers::CASH_NOTE_PURPOSE_OF_TRANSFER;
```

File: sn_client/src/audit/spend_dag.rs [hunk 2]
```diff
@@ -10,12 +10,44 @@ use petgraph::dot::Dot;
 use petgraph::graph::{DiGraph, NodeIndex};
 use petgraph::visit::EdgeRef;
 use serde::{Deserialize, Serialize};
-use sn_transfers::{is_genesis_spend, CashNoteRedemption, NanoTokens, SignedSpend, SpendAddress};
-use std::collections::{BTreeMap, BTreeSet};
-use std::path::Path;
+use sn_transfers::{
+    is_genesis_spend, CashNoteRedemption, NanoTokens, SignedSpend, SpendAddress,
+    CASH_NOTE_PURPOSE_FOR_GENESIS, CASH_NOTE_PURPOSE_FOR_NETWORK_ROYALTIES,
```
Comment (b-zee) [inline sn_client/src/audit/spend_dag.rs:15]:
```suggestion
CASH_NOTE_PURPOSE_OF_GENESIS, CASH_NOTE_PURPOSE_FOR_NETWORK_ROYALTIES,
```

File: sn_client/src/audit/spend_dag.rs [hunk 3]
```diff
@@ -10,12 +10,44 @@ use petgraph::dot::Dot;
...
+
+impl Clone for SpendDagAddress {
+    fn clone(&self) -> Self {
+        SpendDagAddress {
+            address: self.address,
+            purpose: self.purpose.clone(),
+        }
+    }
+}
+
```
Comment (b-zee) [inline sn_client/src/audit/spend_dag.rs]:
```suggestion
#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct SpendDagAddress {
    address: SpendAddress,
    purpose: String,
}
```

File: sn_transfers/src/wallet/hot_wallet.rs [hunk 4]
```diff
@@ -24,7 +24,7 @@ use crate::{
     transfers::{CashNotesAndSecretKey, OfflineTransfer},
     CashNote, CashNoteOutputDetails, CashNoteRedemption, DerivationIndex, DerivedSecretKey, Hash,
     MainPubkey, MainSecretKey, NanoTokens, SignedSpend, Spend, Transaction, Transfer, UniquePubkey,
-    WalletError, NETWORK_ROYALTIES_PK,
+    WalletError, CASH_NOTE_PURPOSE_FOR_NETWORK_ROYALTIES, NETWORK_ROYALTIES_PK,
```
Comment (b-zee) [inline sn_transfers/src/wallet/hot_wallet.rs]:
```suggestion
WalletError, CASH_NOTE_PURPOSE_OF_NETWORK_ROYALTIES, NETWORK_ROYALTIES_PK,
```

File: sn_transfers/src/cashnotes.rs [hunk 5]
```diff
@@ -15,6 +15,11 @@ mod signed_spend;
 mod transaction;
 mod unique_keys;
 
+pub const CASH_NOTE_PURPOSE_FOR_GENESIS: &str = "GENESIS";
+pub const CASH_NOTE_PURPOSE_FOR_NETWORK_ROYALTIES: &str = "ROYALTY";
+pub const CASH_NOTE_PURPOSE_FOR_CHANGE: &str = "CHANGE";
+pub const CASH_NOTE_PURPOSE_FOR_TRANSFER: &str = "TRANSFER";
```
Comment (b-zee) [inline sn_transfers/src/cashnotes.rs]: Better to use `XXX_PURPOSE_OF_XXX`.

File: sn_client/src/audit/spend_dag.rs [hunk 6]
```diff
@@ -10,12 +10,35 @@ use petgraph::dot::Dot;
...
+impl std::fmt::Debug for SpendDagAddress {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "address: {}\npurpose: {}",
+            &self.address.to_hex()[0..6],
+            self.purpose
+        )
+    }
+}
```
Comment (b-zee) [inline sn_client/src/audit/spend_dag.rs:40]:
```suggestion
#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct SpendDagAddress {
    address: SpendAddress,
    purpose: String,
}
```

**Review by b-zee — APPROVED** [review]

File: sn_client/src/audit/spend_dag.rs (hunk 6 above)
Comment (maqi) [reply sn_client/src/audit/spend_dag.rs:40]: the default Debug doesn't have the `line break`, which gives a better output in the diagraph.
hence it has to be individually implemented.

**Review by RolandSherwin — APPROVED** [review]

**Pull request merged**